packages/apputils/src/widgettracker.ts

```diff
@@ -0,0 +1,352 @@
+ * The API surface area of this concrete implementation is substantially larger
+ * than the widget tracker interface it implements. The interface is intended
+ * for export by JupyterLab plugins that create widgets and have clients who may
+ * wish to keep track of newly created widgets. This class, however, can be used
+ * internally by plugins to restore state as well.
+ */
+export class WidgetTracker<T extends Widget = Widget>
+  implements IWidgetTracker<T>, IRestorable<T> {
+  /**
+   * Create a new widget tracker.
+   *
+   * @param options - The instantiation options for a widget tracker.
+   */
+  constructor(options: WidgetTracker.IOptions) {
+    const focus = (this._focusTracker = new FocusTracker());
+    const pool = (this._pool = new RestorablePool(options));
+
+    this.namespace = options.namespace;
+
+    focus.currentChanged.connect((_, current) => {
+      if (current.newValue !== this.currentWidget) {
+        pool.current = current.newValue;
+      }
+    }, this);
+
+    pool.added.connect((_, widget) => {
+      this._widgetAdded.emit(widget);
+    }, this);
+
+    pool.currentChanged.connect((_, widget) => {
+      if (widget === null && focus.currentWidget) {
+        pool.current = focus.currentWidget;
+        return;
+      }
+      this.onCurrentChanged(widget);
+      this._currentChanged.emit(widget);
+    }, this);
+
+    pool.updated.connect((_, widget) => {
+      this._widgetUpdated.emit(widget);
+    }, this);
+  }
+
+  /**
+   * A namespace for all tracked widgets, (e.g., `notebook`).
+   */
+  readonly namespace: string;
+
+  /**
+   * A signal emitted when the current widget changes.
+   */
+  get currentChanged(): ISignal<this, T | null> {
+    return this._currentChanged;
+  }
+
+  /**
+   * The current widget is the most recently focused or added widget.
+   *
+   * #### Notes
+   * It is the most recently focused widget, or the most recently added
+   * widget if no widget has taken focus.
+   */
+  get currentWidget(): T | null {
+    return this._pool.current || null;
+  }
+
+  /**
+   * A promise resolved when the tracker has been restored.
+   */
+  get restored(): Promise<void> {
+    return this._pool.restored;
+  }
+
+  /**
+   * The number of widgets held by the tracker.
+   */
+  get size(): number {
+    return this._pool.size;
+  }
+
+  /**
+   * A signal emitted when a widget is added.
+   *
+   * #### Notes
+   * This signal will only fire when a widget is added to the tracker. It will
+   * not fire if a widget is injected into the tracker.
+   */
+  get widgetAdded(): ISignal<this, T> {
+    return this._widgetAdded;
+  }
+
+  /**
+   * A signal emitted when a widget is updated.
+   */
+  get widgetUpdated(): ISignal<this, T> {
+    return this._widgetUpdated;
+  }
+
+  /**
+   * Add a new widget to the tracker.
+   *
+   * @param widget - The widget being added.
+   */
+  async add(widget: T): Promise<void> {
+    this._focusTracker.add(widget);
+    await this._pool.add(widget);
```

Comment:
In fact, I think we may want to preserve the previous *behavior*, that is to say, that the widget is available in the tracker synchronously after being added, and the promise resolves once it is set in the state DB. That may mean changing this back from an `async` function.

Reply:
For some context, I added that note when we were thinking about `no-floating-promises`, and when it would be safe to use a value inserted into an `InstanceTracker`. We can rethink that behavior, but there *is* a subtle difference here, I think.

Reply:
Even though the function is `async`, when you call it, the widget it still added synchronously. I think I just changed my mind and I agree that adding the note back is a good idea.

For reference, here are two tests that will both succeed:

```typescript
it('should add a widget to the tracker synchronously', async () => {
  const widget = createWidget();
  expect(tracker.has(widget)).to.equal(false);
  void tracker.add(widget);
  expect(tracker.has(widget)).to.equal(true);
  widget.dispose();
});

it('should add a widget to the tracker asynchronously', async () => {
  const widget = createWidget();
  expect(tracker.has(widget)).to.equal(false);
  await tracker.add(widget);
  expect(tracker.has(widget)).to.equal(true);
  widget.dispose();
});
```

Reply:
Ah, didn't see this :)

Reply:
Exactly. Adding the widget synchronously is also more symmetric with the `disposed` behavior, where it is removed synchonously, and then we `void` the connector removal promise.

Reply:
Whenever I have any async function where I rely on some code being execute synchronously before the first `await`, I always leave a comment in the code indicating this. It's useful for other people, and a lot of times, super useful for myself as well:
1. I did this.
2. I meant to do this.
3. This is why.

😅

Reply:
I agree! I think the comment in the docstring accomplishes that.